devito/arguments.py

```diff
@@ -158,198 +139,381 @@ class PtrArgument(Argument):
 
     is_PtrArgument = True
 
-    def __init__(self, name, provider):
-        super(PtrArgument, self).__init__(name, provider, provider.value)
-
-    def verify(self, value):
-        self._value = value or self._value
-        return True
-
-    def reset(self):
-        if callable(self.default_value):
-            self._value = self.default_value()
-        else:
-            self._value = self.default_value
+    def __init__(self, provider):
+        super(PtrArgument, self).__init__(provider.name,
+                                          [Dependency("gets_value_from", provider)],
+                                          provider.dtype)
 
 
-class ArgumentProvider(object):
-
-    """ Abstract base class for any object that, post code-generation, might resolve
-        resolve to runtime arguments. We assume that one source object (e.g. Dimension,
-        SymbolicData) might provide multiple runtime arguments.
+class ArgumentEngine(object):
+    """ Class that encapsulates the argument derivation and verification subsystem
     """
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.stencils = stencils
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
```

Comment:
I hope we can find a way of making the arguments engine unaware of the DLE. These are just arguments in the end...

Reply:
And I don't think we need to store them as instance attributes. They seem to be processed at construction time, and then they also become unnecessary. I may be wrong, I just tried following the logical flow

Reply:
https://github.com/opesci/devito/blob/f9ac081b8f829afc3fde0485f2ed01d2769f46df/devito/arguments.py#L262